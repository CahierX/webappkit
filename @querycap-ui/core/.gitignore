!macro/index.js
!aliases.js