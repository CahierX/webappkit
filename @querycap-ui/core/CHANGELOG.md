# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.4.2](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.4.1...@querycap-ui/core@0.4.2) (2020-03-25)


### Bug Fixes

* dark mode enhance ([10b4685](https://github.com/querycap/webappkit/commit/10b4685758ca6e85379e164dbcacad0df9f99a09))





## [0.4.1](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.4.0...@querycap-ui/core@0.4.1) (2020-03-25)

**Note:** Version bump only for package @querycap-ui/core





# [0.4.0](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.3.1...@querycap-ui/core@0.4.0) (2020-03-25)


### Features

* dark mode ([8836b12](https://github.com/querycap/webappkit/commit/8836b12e58ec24e846bcbc14482d8e67b7c5c5da))





## [0.3.1](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.3.0...@querycap-ui/core@0.3.1) (2020-03-24)


### Bug Fixes

* **form-controls:** styling fix Switch ([16560f1](https://github.com/querycap/webappkit/commit/16560f1865cb19631e6e33957935645cb4abcdef))





# [0.3.0](https://github.com/querycap/webappkit/compare/@querycap-ui/core@0.2.0...@querycap-ui/core@0.3.0) (2020-03-24)


### Features

* **blocks:** add Modal & Popover & Tooltip ([d0cd566](https://github.com/querycap/webappkit/commit/d0cd566f136afe6f29bdab39aeebf55afd04e53d))
* **blocks:** added Dialogs ([bfcc7fe](https://github.com/querycap/webappkit/commit/bfcc7fec2bd8d39d634091780cb3cce7b535d4be))





# 0.2.0 (2020-03-22)


### Features

* **sg:** setup sg ([f612545](https://github.com/querycap/webappkit/commit/f61254522da7bd1bb87c0313036dee2da88cbaf1))
* **ui/core:** rename Theme ([cc8fce5](https://github.com/querycap/webappkit/commit/cc8fce5eea595c2e50c00d0243d2e4c08c3bab9e))
* **uikit:** setup uikit ([7190c50](https://github.com/querycap/webappkit/commit/7190c50a0e5566ea3ef85cba0b1218b4e3d47ba8))





# 0.1.0 (2020-03-13)


### Features

* **ui/core:** rename Theme ([cc8fce5](https://github.com/querycap/webappkit/commit/cc8fce5eea595c2e50c00d0243d2e4c08c3bab9e))
* **uikit:** setup uikit ([7190c50](https://github.com/querycap/webappkit/commit/7190c50a0e5566ea3ef85cba0b1218b4e3d47ba8))
