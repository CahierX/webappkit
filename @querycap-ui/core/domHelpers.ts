export const preventDefault = <T extends { preventDefault: Event["preventDefault"] }>(e?: T) => {
  e && e.preventDefault && e.preventDefault();
  return e;
};

export const stopPropagation = <T extends { stopPropagation: Event["stopPropagation"] }>(e?: T) => {
  e && e.stopPropagation && e.stopPropagation();
  return e;
};

export const stopImmediatePropagation = <T extends { stopImmediatePropagation: Event["stopImmediatePropagation"] }>(
  e?: T,
) => {
  e && e.stopImmediatePropagation && e.stopImmediatePropagation();
  return e;
};
