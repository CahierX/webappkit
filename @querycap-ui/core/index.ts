export * from "polished";
export * from "./theme";
export * from "./select";
export * from "./domHelpers";

export { keyframes } from "@emotion/react";

export * from "react-use-gesture";
export { animated, useSpring, useChain, useTrail, useTransition, useSprings, interpolate, config } from "react-spring";
export type { UseSpringProps, UseTransitionProps, AnimatedValue, UseTransitionResult } from "react-spring";
