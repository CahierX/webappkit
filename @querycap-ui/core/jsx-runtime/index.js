exports.Fragment = require("@emotion/core/jsx-runtime").Fragment;
exports.jsx = require("@emotion/core/jsx-runtime").jsx;
exports.jsxs = require("@emotion/core/jsx-runtime").jsxs;