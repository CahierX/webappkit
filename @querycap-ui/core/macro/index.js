/// @js-check
const { addDefault, addNamed } = require("@babel/helper-module-imports");
const aliases = require("../aliases.json");
const { createMacro } = require("babel-plugin-macros");

const findLast = (arr, predicate) => {
  for (let i = arr.length - 1; i >= 0; i--) {
    if (predicate(arr[i])) {
      return arr[i];
    }
  }
};

const insertBeforeStatements = (program, decl) => {
  const targetPath = findLast(program.get("body"), (p) => p.isImportDeclaration());
  if (targetPath) {
    targetPath.insertAfter([decl]);
  } else {
    program.unshiftContainer("body", decl);
  }
};

module.exports = createMacro(({ references, babel: { types: t } }) => {
  const createImporter = (source) => {
    const exports = {};
    return (path, method) => {
      if (!exports[method]) {
        const program = path.isProgram() ? path : path.findParent((p) => p.isProgram());

        const importDeclaration = program
          .get("body")
          .find((p) => p.isImportDeclaration() && p.node.source.value === source);

        if (importDeclaration) {
          const imported = importDeclaration.node.specifiers.find((n) => n.imported.name === method);
          if (imported) {
            exports[method] = imported.local;
          }
        }

        if (!exports[method]) {
          exports[method] = method === "default" ? addDefault(path, source) : addNamed(path, method, source);
        }
      }
      return exports[method];
    };
  };

  const paramT = "__t";

  const importQuerycapUICore = createImporter("@querycap-ui/core");
  const importEmotionCore = createImporter("@emotion/core");

  const objectExpression = (o) => {
    return t.objectExpression(
      Object.keys(o).map((k) =>
        t.objectProperty(/^[A-Za-z_$][A-Za-z0-9_$]+$/.test(k) ? t.identifier(k) : t.stringLiteral(k), o[k]),
      ),
    );
  };

  const callExpression = (callee, ...args) => {
    return t.callExpression(callee, args);
  };

  const arrowFunctionExpression = (returnExpr, ...params) => {
    return t.arrowFunctionExpression(params, returnExpr);
  };

  const createCSSValues = (selector, blocks) => {
    const v = blocks.length === 1 ? blocks[0] : t.arrayExpression(blocks);
    if (selector) {
      return objectExpression({
        [selector]: v,
      });
    }
    return v;
  };

  const asThemeGetter = (path, arg) => {
    return callExpression(callExpression(importQuerycapUICore(path, "fromTheme"), arg), t.identifier(paramT));
  };

  let styleIdx = 0;

  const createCollector = (path) => {
    const program = path.find((p) => p.isProgram());
    const parts = [];
    let needTheme = false;
    let hasStatic = false;
    let hasThemeGetter = false;
    let cssStaticProps = {};
    let cssThemeGetterProps = {};

    const hoistStaticCSS = (o) => {
      const ident = t.identifier(`__style_${styleIdx}`);
      const cssIdent = importEmotionCore(path, "css");

      const callExpr = callExpression(cssIdent, objectExpression(o));

      const decl = t.variableDeclaration("var", [t.variableDeclarator(ident, callExpr)]);

      insertBeforeStatements(program, decl);

      // to register binding
      if (!program.scope.getBinding(cssIdent.name)) {
        program.traverse({
          ImportSpecifier(s) {
            if (s.get("local").node.name === cssIdent.name) {
              program.scope.registerBinding("module", s);
            }
          },
        });
      }

      // update to referencePaths to let emotion to handle
      // todo find better way
      program.scope.path.traverse({
        Identifier(nodePath) {
          if (nodePath.node.name === ident.name) {
            const callExpr = nodePath.parentPath.get("init");
            program.scope.getBinding(cssIdent.name).referencePaths.push(callExpr.get("callee"));
          }
        },
      });

      styleIdx++;
      return ident;
    };

    const swap = () => {
      if (hasStatic) {
        parts.push(hoistStaticCSS(cssStaticProps));
      }
      if (hasThemeGetter) {
        parts.push(objectExpression(cssThemeGetterProps));
      }
      cssStaticProps = {};
      cssThemeGetterProps = {};
      hasStatic = false;
      hasThemeGetter = false;
    };
    return {
      addWith: (arg) => {
        swap();
        parts.push(asThemeGetter(path, arg));
        needTheme = true;
      },
      addStatic: (key, arg) => {
        const add = (k) => (cssStaticProps[k] = arg);
        if (aliases[key]) {
          aliases[key].forEach(add);
        } else {
          add(key);
        }
        hasStatic = true;
      },
      addThemeGetter: (key, arg) => {
        const add = (k) => (cssThemeGetterProps[k] = asThemeGetter(path, arg));
        if (aliases[key]) {
          aliases[key].forEach(add);
        } else {
          add(key);
        }
        hasThemeGetter = true;
        needTheme = true;
      },
      output: () => {
        swap();
        const selector = path.node.arguments.map((s) => s.value).join(", ");
        if (needTheme) {
          return arrowFunctionExpression(createCSSValues(selector, parts), t.identifier(paramT));
        }
        return createCSSValues(selector, parts);
      },
    };
  };

  const transformSelect = (path) => {
    const collector = createCollector(path);

    const resolveCallChain = (path) => {
      if (path.isCallExpression()) {
        const memberExpression = path.parentPath;
        if (memberExpression.isMemberExpression()) {
          if (memberExpression.parentPath.isCallExpression()) {
            const arg = memberExpression.parentPath.node.arguments[0];
            if (arg) {
              if (memberExpression.node.property.name === "with") {
                collector.addWith(arg);
              } else {
                collector[t.isLiteral(arg) && !t.isTemplateLiteral(arg) ? "addStatic" : "addThemeGetter"](
                  memberExpression.node.property.name,
                  arg,
                );
              }
            }
            return resolveCallChain(memberExpression.parentPath);
          }
        }
        return path;
      }
      return;
    };

    const chainRoot = resolveCallChain(path);

    if (chainRoot) {
      chainRoot.replaceWith(collector.output());
    }
  };

  Object.keys(references).forEach((k) => {
    if (k === "select") {
      references[k].reverse().forEach((p) => transformSelect(p.parentPath));
    } else {
      references[k].reverse().forEach((p) => {
        p.node.name = importQuerycapUICore(p, k).name;
      });
    }
  });
});
