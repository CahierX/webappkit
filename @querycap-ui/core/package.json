{
  "name": "@querycap-ui/core",
  "version": "0.7.4",
  "license": "MIT",
  "peerDependencies": {
    "@babel/runtime-corejs3": "*",
    "lodash": "4.x",
    "react": "16.x"
  },
  "dependencies": {
    "@emotion/core": "10.x",
    "@querycap-ui/core.macro": "0.0.7",
    "@querycap-ui/css-aliases": "0.1.4",
    "polished": "3.5.x",
    "react-spring": "8.0.x"
  },
  "publishConfig": {
    "access": "public"
  },
  "monobundle": {
    "env": "browser",
    "sideDependencies": [
      "@querycap-ui/core.macro"
    ]
  },
  "scripts": {
    "prepare": "monobundle"
  },
  "files": [
    "macro/",
    "aliases.json",
    "dist/"
  ],
  "types": "dist/index.d.ts",
  "main": "dist/index.js",
  "module": "dist/index.es.js"
}
