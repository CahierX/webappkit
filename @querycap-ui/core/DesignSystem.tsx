import { CSSObject, ThemeContext } from "@emotion/core";
import { endsWith } from "lodash";
import React, { ComponentType, ReactNode, useContext, useMemo } from "react";

const color = {
  primary: "#19C7B1",
  secondary: "#27326C",
  success: "#4CD964",
  warning: "#F1911E",
  info: "#A9AEFC",
  danger: "#E51D30",

  text: "#333",
  border: "#e8e8e8",
  bg: "#ffffff",
};

const colorReverse = {
  ...color,
  text: "#FFFFFF",
  border: "#22243B",
  bg: "#000000",
};

export const theme = {
  fontFamily: {
    system: `-apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Helvetica, "PingFang SC", "Hiragino Sans GB", "Microsoft YaHei", SimSun, sans-serif`,
    mono: `"Menlo", "Liberation Mono", "Consolas", "DejaVu Sans Mono", "Ubuntu Mono", "Courier New", "andale mono", "lucida console", monospace`,
  },

  weight: {
    bold: "bold" as CSSObject["fontWeight"],
    normal: "normal" as CSSObject["fontWeight"],
  },

  size: {
    xxs: 10,
    xs: 12,
    s: 14,
    base: 16,
    m: 18,
    l: 24,
    xl: 36,
    xxl: 48,
  },

  color,
  color$$: color,
  colorReverse,
  colorReverse$$: colorReverse,

  radius: {
    s: 2,
    base: 4,
    m: 8,
  },

  spacing: {
    base: "0.6em 1em",
  },

  shadow: {
    dp2: `0 2px 2px 0 rgba(0, 0, 0, .04), 0 3px 1px -2px rgba(0, 0, 0, .2), 0 1px 5px 0 rgba(0, 0, 0, .12)`,
    dp3: `0 3px 4px 0 rgba(0, 0, 0, .04), 0 3px 3px -2px rgba(0, 0, 0, .2), 0 1px 8px 0 rgba(0, 0, 0, .12)`,
    dp4: `0 4px 5px 0 rgba(0, 0, 0, .04), 0 1px 10px 0 rgba(0, 0, 0, .12), 0 2px 4px -1px rgba(0, 0, 0, .2)`,
    dp6: `0 6px 10px 0 rgba(0, 0, 0, .04), 0 1px 18px 0 rgba(0, 0, 0, .12), 0 3px 5px -1px rgba(0, 0, 0, .2)`,
    dp8: `0 8px 10px 1px rgba(0, 0, 0, .04), 0 3px 14px 2px rgba(0, 0, 0, .12), 0 5px 5px -3px rgba(0, 0, 0, .2)`,
    dp16: `0 16px 24px 2px rgba(0, 0, 0, .04), 0 6px 30px 5px rgba(0, 0, 0, .12), 0 8px 10px -5px rgba(0, 0, 0, .2)`,
    dp24: `0 9px 46px 8px rgba(0, 0, 0, .04), 0 11px 15px -7px rgba(0, 0, 0, .12), 0 24px 38px 3px rgba(0, 0, 0, .2)`,
  },
};

function reverse<T>(settings: T): T {
  const nextSettings: any = {};

  for (const k in settings) {
    const val = settings[k];

    if (endsWith(k, "Reverse")) {
      continue;
    }

    if ((settings as any)[`${k}Reverse`]) {
      nextSettings[k] = {
        ...val,
        ...(settings as any)[`${k}Reverse`],
      };

      nextSettings[`${k}Reverse`] = {
        ...(settings as any)[`${k}Reverse`],
        ...val,
      };

      continue;
    }

    nextSettings[k] = val;
  }

  return nextSettings;
}

export type TTheme = typeof theme;

export function ThemeProvider(props: { theme?: TTheme; children: React.ReactNode }) {
  return <ThemeContext.Provider value={props.theme || theme}>{props.children}</ThemeContext.Provider>;
}

export function useDesignSystem(): TTheme {
  // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
  return useContext(ThemeContext) as any;
}

export function DesignSystemColorReverse({ children }: { children?: ReactNode }) {
  const ds = useDesignSystem();
  const r = useMemo(() => reverse(ds), []);
  return <ThemeProvider theme={r}>{children}</ThemeProvider>;
}

export function withDesignSystemReverse<TProps>(Comp: ComponentType<TProps>) {
  return function DesignSystemReverse(props: TProps) {
    return (
      <DesignSystemColorReverse>
        <Comp {...props} />
      </DesignSystemColorReverse>
    );
  };
}
