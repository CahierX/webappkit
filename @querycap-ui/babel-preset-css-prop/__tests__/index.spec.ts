import { transformSync } from "@babel/core";

const compileToSnapshot = (code: string) => {
  const t = transformSync(code, {
    root: __dirname,
    presets: ["@querycap-ui/babel-preset-css-prop"],
  });

  return `
${code}
    
 ↓ ↓ ↓ ↓ ↓ ↓

${t?.code}  
  `;
};

describe("babel-preset-css-prop", () => {
  it("without css prop", () => {
    expect(
      compileToSnapshot(`
import React from "react"

const SomeComponent = () => {
  return (
    <>
      <div>123</div>
    </>
  )
}`),
    ).toMatchSnapshot();
  });

  it("with css prop", () => {
    expect(
      compileToSnapshot(`
import React from "react"

const SomeComponent = () => {
  return (
    <>
      <div css={(t) => ({ color: t.color.primary })}>123</div>
    </>
  )
}`),
    ).toMatchSnapshot();
  });
});
