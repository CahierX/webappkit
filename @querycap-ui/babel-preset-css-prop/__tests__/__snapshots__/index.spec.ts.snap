// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`babel-preset-css-prop with css prop 1`] = `
"

import React from \\"react\\"

const SomeComponent = () => {
  return (
    <>
      <div css={(t) => ({ color: t.color.primary })}>123</div>
    </>
  )
}
    
 ↓ ↓ ↓ ↓ ↓ ↓

import React from \\"react\\";
import { jsx as ___EmotionJSX } from \\"@emotion/core\\";
import { Fragment as ___Fragment } from \\"react\\";

const SomeComponent = () => {
  return ___EmotionJSX(___Fragment, null, ___EmotionJSX(\\"div\\", {
    css: t => ({
      color: t.color.primary
    })
  }, \\"123\\"));
};  
  "
`;

exports[`babel-preset-css-prop without css prop 1`] = `
"

import React from \\"react\\"

const SomeComponent = () => {
  return (
    <>
      <div>123</div>
    </>
  )
}
    
 ↓ ↓ ↓ ↓ ↓ ↓

import React from \\"react\\";

const SomeComponent = () => {
  return React.createElement(React.Fragment, null, React.createElement(\\"div\\", null, \\"123\\"));
};  
  "
`;
